tensorflow/core/kernels/data/sampling_dataset_op.cc

```diff
@@ -0,0 +1,202 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/core/framework/partial_tensor_shape.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/kernels/data/dataset.h"
+#include "tensorflow/core/lib/random/random.h"
+
+namespace tensorflow {
+namespace data {
+namespace {
+
+// See documentation in ../../ops/dataset_ops.cc for a high-level
+// description of the following op.
+
+class SamplingDatasetOp : public UnaryDatasetOpKernel {
+ public:
+  explicit SamplingDatasetOp(OpKernelConstruction* ctx)
+      : UnaryDatasetOpKernel(ctx) {}
+
+ protected:
+  // Create a new SamplingDatasetOp::Dataset, and return it as the output.
+  void MakeDataset(OpKernelContext* ctx, DatasetBase* input,
+                   DatasetBase** output) override {
+    float rate;
+    OP_REQUIRES_OK(ctx, ParseScalarArgument<float>(ctx, "rate", &rate));
+    *output = new Dataset(ctx, rate, input);
+  }
+
+ private:
+  class Dataset : public DatasetBase {
+   public:
+    Dataset(OpKernelContext* ctx, float rate, const DatasetBase* input)
+        : DatasetBase(DatasetContext(ctx)), rate_(rate), input_(input) {
+      input_->Ref();
+    }
+
+    ~Dataset() override { input_->Unref(); }
+
+    std::unique_ptr<IteratorBase> MakeIteratorInternal(
+        const string& prefix) const override {
+      if (rate_ == 0) {
+        return std::unique_ptr<IteratorBase>(new EmptyIterator(
+            {this, strings::StrCat(prefix, "::EmptySample")}));
```

Comment:
s/EmptySample/NoSampling

Reply:
EmptySample had been removed.  The original behavior is endless loop if filter rate is 0.0, by remove EmptySample, the default sample would have exact behavior as the original behavior.